host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,428 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
 }
 ADBG_CASE_DEFINE(pkcs11, 1016, xtest_pkcs11_test_1016,
 		 "PKCS11: Random number generator tests");
+
+static CK_RV derive_sym_key(CK_SESSION_HANDLE session,
+			    CK_OBJECT_HANDLE parent_key,
+			    CK_MECHANISM_TYPE mechanism, size_t data_len,
+			    CK_OBJECT_HANDLE_PTR derv_key_hdl, size_t keylen,
+			    CK_OBJECT_CLASS keyclass, CK_KEY_TYPE keytype,
+			    CK_BBOOL sensitive, CK_BBOOL extble)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	uint8_t buffer[516] = { 0 };
+	CK_MECHANISM mech_derive = { 0 };
+	CK_KEY_DERIVATION_STRING_DATA key_derv_param = { 0 };
+	CK_AES_CBC_ENCRYPT_DATA_PARAMS aes_cbc_param = { 0 };
+	CK_ATTRIBUTE derived_key_template[] = {
+		{ CKA_CLASS, &keyclass, sizeof(keyclass) },
+		{ CKA_KEY_TYPE, &keytype, sizeof(keytype) },
+		{ CKA_ENCRYPT, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_DECRYPT, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_SENSITIVE, &sensitive, sizeof(sensitive) },
+		{ CKA_EXTRACTABLE, &extble, sizeof(extble) },
+		{ CKA_VALUE_LEN, &keylen, sizeof(keylen) }
+	};
+
+	if (data_len > sizeof(buffer))
+		return rv;
+
+	memset(buffer, 0xCC, sizeof(buffer));
+
+	switch (mechanism) {
+	case CKM_AES_ECB_ENCRYPT_DATA:
+		key_derv_param.pData = buffer;
+		key_derv_param.ulLen = data_len;
+		mech_derive.mechanism = mechanism;
+		mech_derive.pParameter = &key_derv_param;
+		mech_derive.ulParameterLen = sizeof(key_derv_param);
+		break;
+	case CKM_AES_CBC_ENCRYPT_DATA:
+		memcpy(aes_cbc_param.iv, buffer, 16);
+		aes_cbc_param.pData = buffer;
```

Comment:
your iv is from data which is probably bad idea -- better change so no-one takes that as an example ;)

Reply:
Sure, I will change that.